### ISSUE TITLE
win.setVisibleOnAllWorkspaces(true) prevents app.dock.hide() from working

### ISSUE THREAD

**Issue by quanglam2807** [electron/electron#25368]
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:** 9.3.0
* **Operating System:** macOS 10.15.6
* **Last Known Working Electron version:** 9.2.1

### Expected Behavior

Dock icon is hidden after calling `app.dock.hide()`. Calling `win.setVisibleOnAllWorkspaces(true)` won't cause the dock icon to resurface.

### Actual Behavior

Dock icon is shown after calling `win.setVisibleOnAllWorkspaces(true)`. 

### To Reproduce

Use the following code:
```js
const { app, BrowserWindow, BrowserView } = require('electron');

app.on('ready', () => {
	app.dock.hide();
	const win = new BrowserWindow({ show: true, frame: false, width: 400, height: 400 });
	win.setVisibleOnAllWorkspaces(true);
});
```
The dock icon will be hidden with Electron 9.2.1 but won't with Electron 9.3.0. 

I've created 3 simple Electron apps for testing:
[9.2.1.zip](https://github.com/electron/electron/files/5188589/9.2.1.zip)  (Electron 9.2.1 with `win.setVisibleOnAllWorkspaces(true)`)
[9.3.0-not-working.zip](https://github.com/electron/electron/files/5188587/9.3.0-not-working.zip) (Electron 9.3.0 with `win.setVisibleOnAllWorkspaces(true)`)
[9.3.0-working.zip](https://github.com/electron/electron/files/5188588/9.3.0-working.zip) (Electron 9.3.0 without `win.setVisibleOnAllWorkspaces(true)`)

### Additional Information

Related to https://github.com/maxogden/menubar/issues/296

Calling `app.dock.hide()` after `win.setVisibleOnAllWorkspaces(true)` is not a good solution. As even though it works, sometimes, the dock icon still resurfaces. I haven't managed to figure out when it happens exactly. Very randomly.

**Comment by quanglam2807** [comment]
So it turns out this is related to https://github.com/electron/electron/pull/25126

It works if I set `visibleOnFullScreen` to `true`
```js
win.setVisibleOnAllWorkspaces(true, { visibleOnFullScreen: true });
```
but doesn't, if I set `visibleOnFullScreen` to `false` (default)
```js
win.setVisibleOnAllWorkspaces(true, { visibleOnFullScreen: false });
```

Is it a bug or is it intentional, @codebytere?

**Comment by taras-slapdash** [comment]
In multi-window applications, another side-effect of `visibleOnFullScreen: true`:

```ts
window.setVisibleOnAllWorkspaces(true, { visibleOnFullScreen: true });
```

is that it hides all other windows when this particular window is open.

@sofianguy, do you have any workaround for this? I really don't want to [downgrade the Electron version again](https://github.com/electron/electron/issues/25360).

**Comment by fanlion** [comment]
> In multi-window applications, another side-effect of `visibleOnFullScreen: true`:
> 
> ```ts
> window.setVisibleOnAllWorkspaces(true, { visibleOnFullScreen: true });
> ```
> 
> is that it hides all other windows when this particular window is open.
> 
> @sofianguy, do you have any workaround for this? I really don't want to [downgrade the Electron version again](https://github.com/electron/electron/issues/25360).

have same issue, in my app, macos

**Comment by dennisfyodorov** [comment]
have same issue, electron v11.0.3

**Comment by jetlej** [comment]
This is a major issue for my users

**Comment by YoucefHQ** [comment]
Same here, painful bug for users.

**Comment by dmamie** [comment]
+1. can we somehow help to get this fixed? :)

**Comment by xinger** [comment]
Electron 11.3.0 — the same bug

> ```js
> win.setVisibleOnAllWorkspaces(true, { visibleOnFullScreen: true });
> ```
This option doesn't help

**Comment by sbannigan** [comment]
Having the same issue on 12.0.6 and 13.0.0-beta.21

**Comment by mathieulj** [comment]
```ts
quickAccessWindow.setVisibleOnAllWorkspaces(true, { visibleOnFullScreen: true, skipTransformProcessType: true });
```
Seems to get this behaving correctly for me.

**Comment by codebytere** [comment]
@mathieulj above is correct - you should call

```
window.setVisibleOnAllWorkspaces(true, { visibleOnFullScreen: true, skipTransformProcessType: true });
``` 

this will not prevent `app.dock.hide()` from working.

**Comment by mathieulj** [comment]
For the record, what I posted is a workaround and finding this issue is how I came to know about it. It feels like at the very least, we need to update the documentation to warn that the API only works as intended (without side effects) when a very specific set of parameters are passed and the correct parameters depend on the dock's visibility state. I hope to put in a documentation PR when I have time to do so. 

It honestly doesn't feel right to just document that avoiding the side effect requires the three parameters to be synchronized according to the state of another api. Unfortunately I'm not well versed enough in objective C and the mac APIs to fully evaluate how difficult it would be to get this working without side effect but [the code comments lead me to believe this might be tricky and/or unattainable](https://github.com/electron/electron/blob/9891ff14a751c98f31aa2e90ddf4bb026d2bbfad/shell/browser/native_window_mac.mm#L1313-L1316).
